specs/tls.rst

```diff
@@ -0,0 +1,201 @@
+==================
+ASGI TLS Extension
+==================
+
+**Version**: 0.2 (2020-09-02)
+
+This specification outlines how to report TLS (or SSL) connection information
+in the ASGI *connection scope* object.
+
+The Base Protocol
+-----------------
+
+TLS is not usable on its own, it always wraps another protocol.
+So this specification is not designed to be usable on its own,
+it must be used as an extension to another ASGI specification.
+That other ASGI specification is referred to as the *base protocol
+specification*.
+
+For HTTP-over-TLS (HTTPS), use this TLS specification and the
+ASGI HTTP specification.  The *base protocol specification* is the
+ASGI HTTP specification.
+
+For WebSockets-over-TLS (wss:// protocol), use this TLS specification
+and the ASGI WebSockets specification.  The *base protocol specification*
+is the ASGI WebSockets specification.
+
+If using this extension with other protocols (not HTTPS or WebSockets), note
+that the *base protocol specification* must define the *connection scope* in a
+way that ensures it covers at most one TLS connection.  If not, you cannot use
+this extension.
+
+When to use this extension
+--------------------------
+
+This extension must only be used for TLS connections.
+
+For non-TLS connections, the ASGI server is forbidden from providing this
+extension.
+
+An ASGI application can check for the presence of the ``"tls"`` extension in
+the ``extensions`` dictionary in the connection scope.  If present, the server
+supports this extension and the connection is over TLS.  If not present,
+either the server does not support this extension or the connection is not
+over TLS.
+
+TLS Connection Scope
+--------------------
+
+The *connection scope* information passed in ``scope`` contains an
+``"extensions"`` key, which contains a dictionary of extensions.  Inside that
+dictionary, the key ``"tls"`` identifies the extension specified in this
+document.  The value will be a dictionary with the following entries:
+
+* ``server_cert`` (*Optional[Unicode string]*) -- The PEM-encoded conversion
+  of the x509 certificate sent by the server when establishing the TLS
+  connection.  Some web server implementations may be unable to provide this
+  (e.g. if TLS is terminated by a separate proxy or load balancer); in that
+  case this shall be ``None``.  Mandatory.
```

Comment:
The way the types are written here is a little confusing if you're not familiar with the typing module - the sentence has both `Optional` and `Mandatory` in it, and the type value is not `Optional[unicode]`. May I suggest the type is instead written as `Unicode string or None` in these cases?

Reply:
Sure, changed.